File: app/data_ingestor.py
"""Module for calculating statistical operations on a CSV"""
import json
import csv

from typing import List, Dict


class TableEntry:
    """Class used to store in memory a row from a CSV table"""

    def __init__(self, index: int, row: list[str]):
        """Initialize a TableEntry instance by providing a row and its index"""
        self.index = index
        self.location_desc = row[4]
        self.question = row[8]
        self.data_value = float(row[11]) if row[11] else None
        self.stratification_category1 = row[30]
        self.stratification1 = row[31]

    def to_json(self) -> str:
        """Returns a JSON-formatted string of the Table_Entry instance"""
        return json.dumps(self.__dict__)

class DataIngestor:
    """
    The DataIngestor reads a CSV file, extracts relevant data
    and stores it in memory as list of TableEntry objects.
    
    It provides methods to perform statistical analyses
    for HTTP POST requests that are made to the server.
    """

    def __init__(self, csv_path: str):
        self.questions_best_is_min = [
            'Percent of adults aged 18 years and older who have an overweight classification',
            'Percent of adults aged 18 years and older who have obesity',
            'Percent of adults who engage in no leisure-time physical activity',
            'Percent of adults who report consuming fruit less than one time daily',
            'Percent of adults who report consuming vegetables less than one time daily'
        ]

        self.questions_best_is_max = [
            'Percent of adults who achieve at least 150 minutes a week '
                + 'of moderate-intensity aerobic physical activity '
                + 'or 75 minutes a week of vigorous-intensity aerobic activity '
                + '(or an equivalent combination)',
            'Percent of adults who achieve at least 150 minutes a week '
                + 'of moderate-intensity aerobic physical activity '
                + 'or 75 minutes a week of vigorous-intensity aerobic physical activity '
                + 'and engage in muscle-strengthening activities on 2 or more days a week',
            'Percent of adults who achieve at least 300 minutes a week '
                + 'of moderate-intensity aerobic physical activity '
                + 'or 150 minutes a week of vigorous-intensity aerobic activity '
                + '(or an equivalent combination)',
            'Percent of adults who engage in muscle-strengthening activities '
                + 'on 2 or more days a week',
        ]


        self.table_entries: List[TableEntry] = []
        self._read_csv(csv_path)

    def _read_csv(self, csv_path):
        """Reads the CSV file and extracts only the relevant columns"""
        with open(csv_path, newline='', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            next(reader)      # Read header row

            for index, row in enumerate(reader):
                try:
                    entry = TableEntry(index, row)
                    self.table_entries.append(entry)
                except ValueError as err:
                    print(f"Skipping row {index} due to error: {err}")


    def compute_response_states_mean(self, question: str) -> Dict:
        """
        Returns the response for '/api/states_mean' request as a JSON dictionary
        """
        selected_rows = list(filter(lambda entry: entry.question == question
                                    and entry.data_value is not None,
                                    self.table_entries))

        state_totals = {}
        state_counts = {}

        for entry in selected_rows:
            state = entry.location_desc
            if state not in state_totals:
                state_totals[state] = 0
                state_counts[state] = 0
            state_totals[state] += entry.data_value
            state_counts[state] += 1

        state_means = {state: total / state_counts[state] for state, total in state_totals.items()}

        sorted_state_means = dict(sorted(state_means.items(), key=lambda item: item[1]))

        return sorted_state_means


    def compute_response_state_mean(self, question: str, state: str) -> Dict:
        """
        Returns the response for '/api/state_mean' request as a JSON dictionary
        """
        selected_rows = list(filter(lambda entry: entry.question == question
                                    and entry.location_desc == state
                                    and entry.data_value is not None,
                                    self.table_entries))

        if not selected_rows:
            return {state: None}

        state_mean = sum(entry.data_value for entry in selected_rows) / len(selected_rows)

        return {state: state_mean}


    def compute_response_best5(self, question: str) -> Dict:
        """
        Returns the response for '/api/best5' request as a JSON dictionary
        """
        selected_rows = list(filter(
            lambda entry: entry.question == question and entry.data_value is not None,
            self.table_entries
        ))

        if not selected_rows:
            return {"error": "No data available for the given question"}

        state_totals = {}
        state_counts = {}

        for entry in selected_rows:
            state = entry.location_desc
            if state not in state_totals:
                state_totals[state] = 0
                state_counts[state] = 0
            state_totals[state] += entry.data_value
            state_counts[state] += 1

        state_means = {state: total / state_counts[state] for state, total in state_totals.items()}

        # Determine sorting order based on question type
        if question in self.questions_best_is_min:
            # Ascending (smallest values are best)
            sorted_states = sorted(state_means.items(), key=lambda item: item[1])
        elif question in self.questions_best_is_max:
            # Descending (largest values are best)
            sorted_states = sorted(state_means.items(), key=lambda item: item[1], reverse=True)
        else:
            return {"error": "Question not found in predefined lists"}

        best5 = dict(sorted_states[:5])

        return best5


    def compute_response_worst5(self, question: str) -> Dict:
        """
        Returns the response for '/api/worst5' request as a JSON dictionary
        """
        selected_rows = list(filter(
            lambda entry: entry.question == question and entry.data_value is not None,
            self.table_entries
        ))

        if not selected_rows:
            return {"error": "No data available for the given question"}

        state_totals = {}
        state_counts = {}

        for entry in selected_rows:
            state = entry.location_desc
            if state not in state_totals:
                state_totals[state] = 0
                state_counts[state] = 0
            state_totals[state] += entry.data_value
            state_counts[state] += 1

        state_means = {state: total / state_counts[state] for state, total in state_totals.items()}

        # Determine sorting order based on question type
        if question in self.questions_best_is_min:
            # Descending (largest values are worst)
            sorted_states = sorted(state_means.items(), key=lambda item: item[1], reverse=True)
        elif question in self.questions_best_is_max:
            # Ascending (smallest values are worst)
            sorted_states = sorted(state_means.items(), key=lambda item: item[1])
        else:
            return {"error": "Question not found in predefined lists"}

        worst5 = dict(sorted_states[:5])
        return worst5

    def compute_response_global_mean(self, question: str) -> Dict:
        """
        Returns the response for '/api/global_mean' request as a JSON dictionary
        """
        selected_values = [
            entry.data_value
            for entry in self.table_entries
            if entry.question == question
            and entry.data_value is not None
        ]

        if not selected_values:
            return {"global_mean": None}

        global_mean = sum(selected_values) / len(selected_values)
        return {"global_mean": global_mean}


    def compute_response_diff_from_mean(self, question: str) -> Dict:
        """
        Returns the response for '/api/diff_from_mean' request as a JSON dictionary
        """
        selected_values = [
            entry.data_value
            for entry in self.table_entries
            if entry.question == question and entry.data_value is not None
        ]

        if not selected_values:
            return {"error": "No data available for the given question"}

        global_mean = sum(selected_values) / len(selected_values)

        state_totals = {}
        state_counts = {}

        for entry in self.table_entries:
            if entry.question == question and entry.data_value is not None:
                state = entry.location_desc
                if state not in state_totals:
                    state_totals[state] = 0
                    state_counts[state] = 0
                state_totals[state] += entry.data_value
                state_counts[state] += 1

        state_means = {
            state: total / state_counts[state] for state, total in state_totals.items()
        }

        diff_from_mean = {state: global_mean - mean for state, mean in state_means.items()}

        sorted_diff_from_mean = dict(
            sorted(
                diff_from_mean.items(),
                key=lambda item: item[1],
                reverse=True
            )
        )

        return sorted_diff_from_mean


    def compute_response_state_diff_from_mean(self, question: str, state: str) -> Dict:
        """
        Returns the response for '/api/state_diff_from_mean' request as a JSON dictionary
        """
        selected_values = [
            entry.data_value
            for entry in self.table_entries
            if entry.question == question
            and entry.data_value is not None
        ]

        if not selected_values:
            return {"error": "No data available for the given question"}

        global_mean = sum(selected_values) / len(selected_values)

        state_values = [
            entry.data_value
            for entry in self.table_entries
            if entry.question == question
            and entry.location_desc == state
            and entry.data_value is not None
        ]

        if not state_values:
            return {state: None}

        state_mean = sum(state_values) / len(state_values)

        diff_from_mean = global_mean - state_mean

        return {state: diff_from_mean}


    def compute_response_mean_by_category(self, question: str) -> Dict:
        """
        Returns the response for '/api/mean_by_category' request as a JSON dictionary
        """
        selected_rows = [
            entry for entry in self.table_entries
            if entry.question == question and entry.data_value is not None
        ]

        # Dacă nu sunt date disponibile, returnăm un JSON vid
        if not selected_rows:
            return {}

        category_totals = {}
        category_counts = {}

        for entry in selected_rows:
            # Exclude entries with empty or invalid stratification categories
            if not entry.stratification_category1 or not entry.stratification1:
                continue

            key = (entry.location_desc, entry.stratification_category1, entry.stratification1)

            if key not in category_totals:
                category_totals[key] = 0
                category_counts[key] = 0

            category_totals[key] += entry.data_value
            category_counts[key] += 1

        # Calculăm media pentru fiecare categorie
        category_means = {
            key: total / category_counts[key] for key, total in category_totals.items()
        }

        # Sorting order
        category_priority = {
            "Age (years)": 1,
            "Education": 2,
            "Gender": 3,
            "Income": 4,
            "Race/Ethnicity": 5,
            "Total": 6
        }

        age_priority = {
            "18 - 24": 1, "25 - 34": 2, "35 - 44": 3, "45 - 54": 4, "55 - 64": 5, "65 or older": 6
        }

        education_priority = {
            "Less than high school": 1, "High school graduate": 2,
            "Some college or technical school": 3, "College graduate": 4
        }

        income_priority = {
            "Less than $15,000": 1, "$15,000 - $24,999": 2, "$25,000 - $34,999": 3,
            "$35,000 - $49,999": 4, "$50,000 - $74,999": 5, "$75,000 or greater": 6,
            "Data not reported": 7
        }

        def custom_sort_key(item):
            state, category, value = item[0]
            category_rank = category_priority.get(category, 99)

            if category == "Age (years)":
                value_rank = age_priority.get(value, 99)
            elif category == "Education":
                value_rank = education_priority.get(value, 99)
            elif category == "Income":
                value_rank = income_priority.get(value, 99)
            else:
                # Default for categories without custom ranking
                value_rank = 99

            return (state, category_rank, value_rank, value)

        sorted_means = dict(sorted(category_means.items(), key=custom_sort_key))

        return {str(k): v for k, v in sorted_means.items()}



    def compute_response_state_mean_by_category(self, question: str, state: str) -> Dict:
        """
        Returns the response for '/api/state_mean_by_category' request as a JSON dictionary
        """
        # Select rows that match the question and the state, with valid data_value
        selected_rows = [
            entry for entry in self.table_entries
            if entry.question == question
                and entry.location_desc == state
                and entry.data_value is not None
        ]

        if not selected_rows:
            return {"error": "No data available for the given question and state"}

        # Dictionary to store the total values and counts
        # for each stratification category and stratification
        category_totals = {}
        category_counts = {}

        for entry in selected_rows:
            key = (entry.stratification_category1, entry.stratification1)

            if key not in category_totals:
                category_totals[key] = 0
                category_counts[key] = 0

            category_totals[key] += entry.data_value
            category_counts[key] += 1

        # Calculate the mean for each category-stratification combination
        category_means = {
            key: total / category_counts[key]
            for key, total in category_totals.items()
        }

        calculated_values = {str(k): v for k, v in category_means.items()}
        return {state: calculated_values}

File: method-rs/src/data_ingestor.rs
use csv::ReaderBuilder;
use serde::{Deserialize, Serialize};
use serde_json::{self, json};
use std::{
    collections::{self, HashMap, btree_map::Range},
    fs::File,
    os::linux::raw::stat,
};

/// This is best-effort (cannot store String as const)

const QUESTIONS_BEST_IS_MIN: &[&str] = &[
    "Percent of adults aged 18 years and older who have an overweight classification",
    "Percent of adults aged 18 years and older who have obesity",
    "Percent of adults who engage in no leisure-time physical activity",
    "Percent of adults who report consuming fruit less than one time daily",
    "Percent of adults who report consuming vegetables less than one time daily",
];

const QUESTIONS_BEST_IS_MAX: &[&str] = &[
    "Percent of adults who achieve at least 150 minutes a week \
        of moderate-intensity aerobic physical activity \
        or 75 minutes a week of vigorous-intensity aerobic activity \
        (or an equivalent combination)",
    "Percent of adults who achieve at least 150 minutes a week \
        of moderate-intensity aerobic physical activity \
        or 75 minutes a week of vigorous-intensity aerobic physical activity \
        and engage in muscle-strengthening activities on 2 or more days a week",
    "Percent of adults who achieve at least 300 minutes a week \
        of moderate-intensity aerobic physical activity \
        or 150 minutes a week of vigorous-intensity aerobic activity \
        (or an equivalent combination)",
    "Percent of adults who engage in muscle-strengthening activities \
        on 2 or more days a week",
];

#[derive(PartialEq)]
enum RankingType {
    Best5,
    Worst5,
}

// Relevant columns from CSV file
#[derive(Debug, Serialize)]
struct TableEntry {
    index: u32,
    location_desc: String,
    question: String,
    data_value: f32,
    stratification_category1: String,
    stratification1: String,
}

#[derive(Debug, Serialize)]
pub struct Table {
    pub entries: Vec<TableEntry>,
}

pub fn load_csv(filename: &str) -> Table {
    let file = File::open(filename).expect("Cannot open file");

    let mut rdr = csv::Reader::from_reader(file);
    let headers = rdr
        .headers()
        .expect("CSV file doesn't have header line")
        .clone();

    let mut entries: Vec<TableEntry> = vec![];
    let mut line_index = 0;

    for result in rdr.records() {
        line_index = line_index + 1;
        match result {
            Ok(entry) => {
                let row: HashMap<_, _> = headers.iter().zip(entry.iter()).collect();

                // Get cell value by column name:

                let location_desc: &str = row
                    .get("LocationDesc")
                    .expect("Missing \"LocationDesc\" column");

                let question: &str = row.get("Question").expect("Missing \"Question\" column");

                let data_value: f32 = row
                    .get("Data_Value")
                    .expect("Missing \"Data_Value\" column")
                    .parse::<f32>()
                    .expect("Cannot parse \"Data_Value\" as f32");

                let stratification_category1: &str = row
                    .get("StratificationCategory1")
                    .expect("Missing \"StratificationCategory1\" column");

                let stratification1: &str = row
                    .get("Stratification1")
                    .expect("Missing \"Stratification1\" column");

                entries.push(TableEntry {
                    index: line_index,
                    location_desc: location_desc.to_string(),
                    question: question.to_string(),
                    data_value: data_value,
                    stratification_category1: stratification_category1.to_string(),
                    stratification1: stratification1.to_string(),
                })
            }
            Err(err) => eprintln!("[ERR] Skipping row due to: {}", err),
        }
    }

    Table { entries: entries }
}

fn compute_states_mean<'a>(table: &'a Table, question: &'a str) -> HashMap<&'a str, f32> {
    let selected_entries: Vec<&TableEntry> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question)
        .collect();

    let mut state_totals: HashMap<&str, f32> = HashMap::new();
    let mut state_counts: HashMap<&str, u32> = HashMap::new();

    for entry in selected_entries {
        let state: &str = &entry.location_desc;

        if let Some(value) = state_totals.get(state) {
            // HashMap contains key
            state_totals.insert(&state, value + entry.data_value);
        } else {
            // HashMap does NOT contain key
            state_totals.insert(&state, 0.0);
        }

        if let Some(value) = state_counts.get(state) {
            // HashMap contains key
            state_counts.insert(&state, value + 1);
        } else {
            // HashMap does NOT contain key
            state_counts.insert(&state, 0);
        }
    }

    let mut states_mean: HashMap<&str, f32> = HashMap::new();

    for (state, total) in state_totals.iter() {
        states_mean.insert(
            &state,
            total / state_counts.get(state).unwrap().clone() as f32,
        );
    }

    states_mean
}

pub fn json_response_states_mean(table: &Table, question: &str) -> String {
    // TODO:
    let states_mean: HashMap<&str, f32> = compute_states_mean(table, question);
    serde_json::to_string(&states_mean).unwrap()
}

pub fn json_response_state_mean(table: &Table, question: &str, state: &str) -> String {
    // TODO:
    let selected_entries: Vec<&TableEntry> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question && entry.location_desc == state)
        .collect();

    if selected_entries.is_empty() {
        // TODO: treat this case
        return String::new();
    }

    let state_total: f32 = selected_entries.iter().map(|entry| entry.data_value).sum();

    let mut state_mean: HashMap<&str, f32> = HashMap::new();

    state_mean.insert(&state, state_total / selected_entries.len() as f32);

    serde_json::to_string(&state_mean).unwrap()
}

fn get_best5_or_worst5(
    states_mean: &HashMap<&str, f32>,
    question: &str,
    ranking_type: RankingType,
) -> String {
    let is_question_in_best_mins: bool = QUESTIONS_BEST_IS_MIN.contains(&question);
    let is_question_in_best_maxs: bool = QUESTIONS_BEST_IS_MAX.contains(&question);

    if (is_question_in_best_mins == true && ranking_type == RankingType::Best5)
        || (is_question_in_best_maxs == true && ranking_type == RankingType::Worst5)
    {
        // Sort in ASCENDING order (smaller values are best)
        let mut sorted: Vec<(&&str, &f32)> = states_mean.iter().collect();
        sorted.sort_by(|a, b| a.1.partial_cmp(b.1).unwrap());

        // Take first 5 (handles cases with <5 elements)
        let top5_vec = &sorted[..5.min(sorted.len())];

        // Convert to HashMap
        let top5_map: HashMap<_, _> = top5_vec.iter().map(|(k, v)| (k, **v)).collect();

        serde_json::to_string(&top5_map).unwrap()
    } else if (is_question_in_best_maxs == true && ranking_type == RankingType::Best5)
        || (is_question_in_best_mins == true && ranking_type == RankingType::Worst5)
    {
        // Sort in DESCENDING order (bigger values are best)
        let mut sorted: Vec<(&&str, &f32)> = states_mean.iter().collect();
        sorted.sort_by(|a, b| b.1.partial_cmp(a.1).unwrap());

        // Take first 5 (handles cases with <5 elements)
        let top5_vec = &sorted[..5.min(sorted.len())];

        // Convert to HashMap
        let top5_map: HashMap<_, _> = top5_vec.iter().map(|(k, v)| (k, **v)).collect();

        serde_json::to_string(&top5_map).unwrap()
    } else {
        let mut response: HashMap<&str, &str> = HashMap::new();
        response.insert("error", "Question not found in predefined lists");

        serde_json::to_string(&response).unwrap()
    }
}

pub fn json_response_best5(table: &Table, question: &str) -> String {
    // TODO
    let states_mean: HashMap<&str, f32> = compute_states_mean(table, question);

    if states_mean.is_empty() {
        let mut response: HashMap<&str, &str> = HashMap::new();
        response.insert(&"error", &"No data available for the given question");

        serde_json::to_string(&response).unwrap()
    } else {
        get_best5_or_worst5(&states_mean, question, RankingType::Best5)
    }
}

pub fn json_response_worst5(table: &Table, question: &str) -> String {
    // TODO
    let states_mean: HashMap<&str, f32> = compute_states_mean(table, question);

    if states_mean.is_empty() {
        let mut response: HashMap<String, String> = HashMap::new();
        response.insert(
            "error".to_string(),
            "No data available for the given question".to_string(),
        );

        serde_json::to_string(&response).unwrap()
    } else {
        get_best5_or_worst5(&states_mean, question, RankingType::Worst5)
    }
}

pub fn json_response_global_mean(table: &Table, question: &str) -> String {
    let selected_values: Vec<f32> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question)
        .map(|entry| entry.data_value)
        .collect();

    if selected_values.is_empty() {
        let mut response: HashMap<&str, Option<()>> = HashMap::new();
        response.insert(&"global_mean", None);

        serde_json::to_string(&response).unwrap()
    } else {
        let sum: f32 = selected_values.iter().sum();
        let global_mean: f32 = sum / selected_values.len() as f32;
        let mut response: HashMap<&str, f32> = HashMap::new();
        response.insert(&"global_mean", global_mean);

        serde_json::to_string(&response).unwrap()
    }
}

pub fn json_response_diff_from_mean(table: &Table, question: &str) -> String {
    let selected_values: Vec<f32> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question)
        .map(|entry| entry.data_value)
        .collect();

    if selected_values.is_empty() {
        let sum: f32 = selected_values.iter().sum();
        let global_mean: f32 = sum / selected_values.len() as f32;
        let mut response: HashMap<&str, f32> = HashMap::new();
        response.insert(&"global_mean", global_mean);

        return serde_json::to_string(&response).unwrap();
    }

    // Using ::<f32> turbofish
    let global_mean: f32 = selected_values.iter().sum::<f32>() / selected_values.len() as f32;

    let states_mean: HashMap<&str, f32> = compute_states_mean(table, question);

    let mut diff_from_mean: HashMap<&str, f32> = HashMap::new();

    for (state, mean) in states_mean.iter() {
        diff_from_mean.insert(&state, global_mean - mean);
    }

    // Operations: HashMap -> Vec -> sort DESC -> HashMap

    // Sort in DESCENDING order
    let mut sorted: Vec<(&str, f32)> = diff_from_mean.into_iter().collect();

    sorted.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());

    let diff_from_mean: HashMap<&str, f32> = sorted.into_iter().map(|(k, v)| (k, v)).collect();

    serde_json::to_string(&diff_from_mean).unwrap()
}

fn get_category_priorities<'a>() -> HashMap<&'a str, u32> {
    let mut category_priorities: HashMap<&str, u32> = HashMap::new();
    category_priorities.insert("Age (years)", 1);
    category_priorities.insert("Education", 2);
    category_priorities.insert("Gender", 3);
    category_priorities.insert("Income", 4);
    category_priorities.insert("Race/Ethnicity", 5);
    category_priorities.insert("Total", 6);

    category_priorities
}

fn get_age_priorities<'a>() -> HashMap<&'a str, u32> {
    let mut age_priorities: HashMap<&str, u32> = HashMap::new();
    age_priorities.insert("18 - 24", 1);
    age_priorities.insert("25 - 34", 2);
    age_priorities.insert("35 - 44", 3);
    age_priorities.insert("45 - 54", 4);
    age_priorities.insert("55 - 64", 5);
    age_priorities.insert("65 or older", 6);

    age_priorities
}

fn get_education_priorities<'a>() -> HashMap<&'a str, u32> {
    let mut education_priorities: HashMap<&str, u32> = HashMap::new();
    education_priorities.insert("Less than high school", 1);
    education_priorities.insert("High school graduate", 2);
    education_priorities.insert("Some college or technical school", 3);
    education_priorities.insert("College graduate", 4);

    education_priorities
}

fn get_income_priorities<'a>() -> HashMap<&'a str, u32> {
    let mut income_priorities: HashMap<&str, u32> = HashMap::new();
    income_priorities.insert("Less than $15,000", 1);
    income_priorities.insert("$15,000 - $24,999", 2);
    income_priorities.insert("$25,000 - $34,999", 3);
    income_priorities.insert("$35,000 - $49,999", 4);
    income_priorities.insert("$50,000 - $74,999", 5);
    income_priorities.insert("$75,000 or greater", 6);
    income_priorities.insert("Data not reported", 7);

    income_priorities
}

pub fn json_response_mean_by_catagory(table: &Table, question: &str) -> String {
    let selected_entries: Vec<&TableEntry> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question)
        .collect();

    if selected_entries.is_empty() {
        return serde_json::to_string("").unwrap();
    }

    let mut category_totals: HashMap<(&str, &str, &str), f32> = HashMap::new();
    let mut category_counts: HashMap<(&str, &str, &str), u32> = HashMap::new();

    for entry in selected_entries.iter() {
        if entry.stratification1 == "" || entry.stratification_category1 == "" {
            continue;
        }

        let key: (&str, &str, &str) = (
            &entry.location_desc,
            &entry.stratification_category1,
            &entry.stratification1,
        );

        if let Some(value) = category_totals.get(&key) {
            // HashMap contains key
            category_totals.insert(key.clone(), value + entry.data_value);
        } else {
            // HashMap does NOT contain key
            category_totals.insert(key.clone(), entry.data_value);
        }

        if let Some(value) = category_counts.get(&key) {
            // HashMap contains key
            category_counts.insert(key.clone(), value + 1);
        } else {
            // HashMap does NOT contain key
            category_counts.insert(key.clone(), 1);
        }
    }

    let mut category_means: HashMap<(&str, &str, &str), f32> = HashMap::new();

    for (key, total) in category_totals.iter() {
        category_means.insert(*key, total / *category_counts.get(key).unwrap() as f32);
    }

    let category_priorities: HashMap<&str, u32> = get_category_priorities();
    let age_priorities: HashMap<&str, u32> = get_age_priorities();
    let education_priorities: HashMap<&str, u32> = get_education_priorities();
    let income_priorities: HashMap<&str, u32> = get_income_priorities();

    let mut sorted: Vec<((&str, &str, &str), f32)> = category_means.into_iter().collect();

    // TODO: continue it

    String::new()
}

pub fn json_response_state_mean_by_category(table: &Table, question: &str, state: &str) -> String {
    let selected_entries: Vec<&TableEntry> = table
        .entries
        .iter()
        .filter(|entry| entry.question == question && entry.location_desc == state)
        .collect();

    if selected_entries.is_empty() {
        let mut response: HashMap<&str, &str> = HashMap::new();
        response.insert(
            "error",
            "No data avialable for the given question and state",
        );

        return serde_json::to_string(&response).unwrap();
    }

    let mut category_totals: HashMap<(&str, &str), f32> = HashMap::new();
    let mut category_counts: HashMap<(&str, &str), u32> = HashMap::new();

    for entry in selected_entries {
        let key: (&str, &str) = (&entry.stratification_category1, &entry.stratification1);

        if let Some(value) = category_totals.get(&key) {
            // HashMap contains key
            category_totals.insert(key.clone(), value + entry.data_value);
        } else {
            // HashMap does NOT contain key
            category_totals.insert(key.clone(), entry.data_value);
        }

        if let Some(value) = category_counts.get(&key) {
            // HashMap contains key
            category_counts.insert(key.clone(), value + 1);
        } else {
            // HashMap does NOT contain key
            category_counts.insert(key.clone(), 1);
        }
    }

    let mut category_means: HashMap<(&str, &str), f32> = HashMap::new();
    for (key, total) in category_totals.iter() {
        category_means.insert(*key, total / *category_counts.get(key).unwrap() as f32);
    }

    // TODO: continue

    serde_json::to_string(&category_means).unwrap()
}
